Splitting functional tests as cloud and non-cloud

ytimocin opened the pull request:
# Description

The goal of this PR is to create another workflow to run functional tests that don't use cloud resources and also remove the non-cloud tests from the existing functional test workflow that requires approval. In brief:

1. Workflow: **functional-test-cloud** (This requires an approval)
2. Workflow: **functional-test-noncloud** (This doesn't require an approval and it should be enabled on forks)

**RP tests that use cloud resources:**
* Core
* DataStores
* UCP

**RP tests that use non-cloud resources:**
* CLI
* Core
* Dapr
* DataStores
* Kubernetes
* Messaging
* Samples
* UCP

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
Fixes: #7624

ytimocin committed:
Splitting functional tests as cloud and non-cloud

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,697 @@
+          which rad || { echo "cannot find rad"; exit 1; }
+
+          echo "*** Installing Radius to Kubernetes ***"
+          rad install kubernetes \
```
Getting this error: `*** Installing Radius to Kubernetes ***
Installing Radius version afefa5b to namespace: radius-system...
Error: failed to run Radius Helm install, err: 
cannot re-use a name that is still in use
Helm output:
getting history for release radius
creating 1 resource(s)
creating 1 resource(s)
creating 1 resource(s)
beginning wait for 3 resources with timeout of 1m0s`.

Link to an example run: https://github.com/radius-project/radius/actions/runs/9659072267/job/26641840926.

ytimocin replied:
Added `--reinstall` flag and retesting: https://github.com/radius-project/radius/actions/runs/9662144721.

ytimocin replied:
Getting the same error.

ytimocin replied:
Removing `--reinstall`.

willdavsmith commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,741 @@
+        id: get_installation_token
+        with:
+          app_id: ${{ env.FUNCTIONAL_TEST_APP_ID }}
+          private_key: ${{ secrets.FUNCTIONAL_TEST_APP_PRIVATE_KEY }}
```
is there any risk of exposing secrets since these can run without approval?

ytimocin committed:
Fixing a format issue

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/mongodb_test.go:
```diff
@@ -70,49 +70,3 @@ func Test_MongoDB_Manual(t *testing.T) {
-
-// Test_MongoDB_Recipe validates:
-// the creation of a mongoDB from a recipe that uses an Azure resource
-func Test_MongoDB_Recipe(t *testing.T) {
```
Moved this test to the **cloud** package.

rynowak replied:
Thank you! 🙇

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/redis_test.go:
```diff
@@ -72,75 +69,3 @@ func Test_Redis_Manual(t *testing.T) {
 	test.Test(t)
 }
-
-func Test_Redis_Recipe(t *testing.T) {
```
Moved these tests to the **cloud** package.

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep:
This wasn't being used anywhere.

ytimocin replied:
Or maybe it was 🤨

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-redis-manual.bicep:
Refactor was necessary.

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-simulatedenv-recipe.bicep:
Refactor was necessary.

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-sqldb-manual.bicep:
Refactor was necessary.

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-rs-mongodb-manual.bicep:
```diff
@@ -35,8 +35,6 @@ resource webapp 'Applications.Core/containers@2023-10-01-preview' = {
 }
 
-
-// https://hub.docker.com/_/mongo/
```
Deleted this comment as I think it is unnecessary.

ytimocin commented on .github/workflows/functional-test.yaml:
```diff
@@ -33,31 +34,31 @@ on:
 
   # gotestsum version - see: https://github.com/gotestyourself/gotestsum
-  GOTESTSUM_VER: 1.10.0
+  GOTESTSUM_VER: 1.12.0
```
It's been more than a year since the 1.10.0 version was released so I updated it: https://github.com/gotestyourself/gotestsum/releases.

ytimocin commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,461 @@
+          echo "PR_NUMBER=${{ env.PR_NUMBER }}" >> $GITHUB_OUTPUT
+          echo "DE_IMAGE=${{ env.DE_IMAGE }}" >> $GITHUB_OUTPUT
+          echo "DE_TAG=${{ env.DE_TAG }}" >> $GITHUB_OUTPUT
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
```
If this is being run from a fork, then logging into GHCR should be with that user's credentials.

ytimocin commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,461 @@
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and Push container images
+        run: |
+          make build && make docker-build && make docker-push
```
If this is from a fork, should these be published to the user's packages?

ytimocin commented on .github/actions/process-test-results/action.yaml:
Reformatting all the files that I touch during the development of this PR.

lakshmimsft commented on .github/workflows/functional-test.yaml:
```diff
@@ -14,12 +14,13 @@
 # ------------------------------------------------------------
 
-name: Functional tests
+name: Functional Tests (with Cloud Resources)
```
Could we rename this file to functional-test-cloud.yaml?

lakshmimsft commented on test/functional-portable/datastoresrp/cloud/resources/redis_test.go:
```diff
@@ -0,0 +1,101 @@
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Redis_Recipe(t *testing.T) {
```
I'm missing something. why is this moved to cloud resources tests?

ytimocin replied:
It is using an Azure provider in the environment. Please see: https://github.com/radius-project/radius/blob/b961aa02f67375eb5e809559cfad77f8d5509c6f/test/functional-portable/datastoresrp/cloud/resources/testdata/datastoresrp-resources-redis-recipe.bicep.

rynowak replied:
Why though? Does it actually require an Azure resource? 

We should have a justification when tests require the cloud.

ytimocin commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,461 @@
+  # Kubectl version
+  KUBECTL_VER: "v1.25.0"
+  # Container registry for storing container images
+  CONTAINER_REGISTRY: ghcr.io/${{ github.repository_owner }}/dev
```
We can run a registry in the Kubernetes cluster.

ytimocin replied:
Add `radius` in the URL.

lakshmimsft commented:
The Approval workflow today triggers ensures all tests are run. Since we're splitting functional-tests into cloud and non-cloud, will that be updated? to run only cloud tests or no?
How do we ensure in the future that both cloud and non-cloud tests have be run?

ytimocin commented:
> The Approval workflow today triggers ensures all tests are run. Since we're splitting functional-tests into cloud and non-cloud, will that be updated? to run only cloud tests or no? How do we ensure in the future that both cloud and non-cloud tests have be run?

When someone creates a PR the functional tests with non-cloud resources will run by default. The other one, functional tests with cloud resources, will need an approval like it does today. Basically, all the tests will have run. Here is where the one with the approval is: https://github.com/radius-project/radius/pull/7716/files#diff-a5e3050f5c1a812d9565ba0c87fb5028ad0a004c2d966e71d67208b4b33042c4L327-R332.

ytimocin commented on build/recipes.mk:
```diff
@@ -15,7 +15,6 @@
 # ------------------------------------------------------------
 
 BICEP_RECIPE_TAG_VERSION?=latest
-RAD_BICEP_PATH?=$${HOME}/.rad/bin
```
This is not being used anywhere in the code. I did a search in the whole project. Please correct me if I am wrong.

rynowak replied:
If it's unused then it's good for us to delete !

ytimocin commented on test/functional-portable/datastoresrp/noncloud/resources/testdata/datastoresrp-resources-mongodb-recipe-context.bicep:
This wasn't being used.

ytimocin committed:
Reverting back to local registry

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Adding a secure local registry

Signed-off-by: ytimocin <ytimocin@microsoft.com>

rynowak commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,445 @@
+    outputs:
+      REL_VERSION: ${{ steps.gen-id.outputs.REL_VERSION }}
+      DE_IMAGE: ${{ steps.gen-id.outputs.DE_IMAGE }}
+      DE_TAG: ${{ steps.gen-id.outputs.DE_TAG }}
```
Where do these get set? `gen-id` runs as part of this job, so how can it provide the outputs?

rynowak commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,445 @@
+          echo "CHECKOUT_REPO=${{ github.repository }}" >> $GITHUB_ENV
+          echo "CHECKOUT_REF=${{ github.ref }}" >> $GITHUB_ENV
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+
```
Don't we also need to setup the checkout target for a workflow dispatch here?

rynowak commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,445 @@
+          echo "CHECKOUT_REF=${{ github.ref }}" >> $GITHUB_ENV
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+
+      - name: Use custom actions
```
This doesn't sound right 😆

rynowak commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,445 @@
+        env:
+          DOCKER_REGISTRY: "${{ env.LOCAL_REGISTRY_SERVER }}:${{ env.LOCAL_REGISTRY_PORT }}"
+          DOCKER_TAG_VERSION: ${{ env.REL_VERSION }}
+
```
Very cool to see this come together!

rynowak commented on .github/workflows/functional-test-noncloud.yaml:
```diff
@@ -0,0 +1,445 @@
+      - uses: azure/setup-kubectl@v4
+        if: always()
+        with:
+          version: ${{ env.KUBECTL_VER }}
```
Do we need this here? I think the scripts already called into kubectl. IMO it's better to install dependencies early in the workflow instead of near the end.

ytimocin replied:
I removed this step from both workflows.

rynowak commented on test/functional-portable/cli/noncloud/cli_test.go:
```diff
@@ -68,7 +68,9 @@ func verifyRecipeCLI(ctx context.Context, t *testing.T, test rp.RPTest) {
-	target := fmt.Sprintf("br:ghcr.io/radius-project/dev/test-bicep-recipes/redis-recipe:%s", generateUniqueTag())
+
+	// The target is a local registry, so we can use localhost.
+	target := fmt.Sprintf("br:localhost:5000/dev/test-bicep-recipes/redis-recipe:%s", generateUniqueTag())
```
Should we pass this in via an env-var since this requires additional setup? Then we could skip if the env-var is not set.

It's best if the functional tests "just work" without additional steps like a local registry.

rynowak commented on test/functional-portable/datastoresrp/cloud/resources/redis_test.go:
Do we have coverage of these resources without cloud resources? I thought all of the datastores RP code was tested without a cloud dependency..

rynowak reviewed (changes requested):
Great to see the progress on this!

I think my only major concern is about the tests that are being moved to the cloud package. Can we huddle quickly and figure out why this is needed?

We should use cloud resources iff we need to test cloud features.

ytimocin committed:
Addressing comments

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented on .github/actions/create-kind-cluster/action.yaml:
```diff
@@ -0,0 +1,110 @@
+          skip_verify = true
+        EOF
+
+          # RADIUS_DIR
```
Removed

ytimocin commented on .github/actions/create-kind-cluster/action.yaml:
```diff
@@ -0,0 +1,110 @@
+      if: ${{ inputs.with-local-registry == 'true' }}
+      shell: bash
+      run: |
+        cat <<EOF | kind create cluster --config=-
```
Yes, and added the reference.

ytimocin commented on .github/actions/create-kind-cluster/action.yaml:
```diff
@@ -0,0 +1,110 @@
+  registry-port:
+    description: "The port for the local registry"
+    required: false
+    default: "5000"
```
We don't need any output for now. The callee has context of the URL and the registry name because they can be changed by the callee.

ytimocin commented on .github/actions/create-local-registry/action.yaml:
```diff
@@ -0,0 +1,110 @@
+    description: "The port for the local registry"
+    required: false
+    default: "5000"
+runs:
```
Same comment as above.

ytimocin commented:
https://github.com/radius-project/radius/pull/7716#discussion_r1683629719 There were some bicep files that used Azure provider for the environment and that is why I had moved those tests to the cloud package in the datastoresrp functional test. Then I checked and saw that we don't need the provider. I removed those parts and moved all those tests back to the noncloud package. cc/ @rynowak

rynowak commented on .github/actions/create-kind-cluster/action.yaml:
```diff
@@ -0,0 +1,110 @@
+        - role: control-plane
+          extraMounts:
+            - containerPath: "/etc/containerd/certs.d/${{ inputs.registry-name }}"
+              hostPath: "$TEMP_CERT_DIR/certs/${{ inputs.registry-server }}"
```
Having this environment variable set is a prerequisite for using this action right? Wouldn't that be better as a parameter since it's required?

ytimocin replied:
Oh, I see now. I will update this.

rynowak replied:
This is still using the environment variable. Can you update it please?

rynowak commented on test/testutil/testutil.go:
```diff
@@ -335,3 +335,12 @@ func GetCreationTimestamp() string {
 	return "radiusfunctionaltestbucket-" + uuid.New().String()
 }
+
+func GetDockerRegistryName() string {
```
How is this different from `DOCKER_REGISTRY`? Looks like they are two different settings with the same default value. Maybe a comment would help?

ytimocin replied:
I added a comment. In the case of using a local registry, DOCKER_REGISTRY will return the port too, so I added this function that returns the name of the registry without a port if present.

rynowak commented on .github/actions/create-kind-cluster/action.yaml:
```diff
@@ -0,0 +1,148 @@
+    description: "Whether the KinD cluster should be created with a secure local registry configuration"
+    required: false
+    default: "false"
+  temp-cert-dir:
```
suggestion: `cert-dir`. Saying "temp" doesn't really make sense as a parameter name. I think it's a translation from the original code.

sylvainsf reviewed (approved):
Looks great!
